### ISSUE TITLE
fix: noderesources plugin flaw

### ISSUE THREAD

**Issue by LY-today** [kubernetes/kubernetes#129229]
### What happened?

The score function of resourceAllocationScorer should not iterate over args.ScoringStrategy.Resources. For example: the strategy parameters are set to [gpu:2,cpu:1,mem:1]. At this time, the pod only applies for cpu and mem, but because k8s traverses parameters instead of traversing pod application, gpu becomes the final The key factor in scoring, but in fact, the pod only needs to consider resource application issues of CPU and MEM, and does not need to consider GPU. After all, it has not applied for GPU. If it is the MostRequestedPriority policy at this time, after consideration, the CPU strength will be scheduled to the GPU. On the machine, there is a fragmentation problem

### What did you expect to happen?

Which resources will be considered for scoring based on which resources the pod has applied for,Instead of scoring all the resources configured in the configuration file, which may be counterproductive in real scenarios.

### How can we reproduce it (as minimally and precisely as possible)?

The MostRequestedPriority policy parameters are set to [gpu:2,cpu:1,mem:1]. At this time, the pod only applies for cpu and mem. It is found that the pod is scheduled to a machine with concentrated GPUs. As a result, when there are large model GPU tasks that need to be scheduled, The machine's GPU is satisfied but the CPU is not.

### Kubernetes version

latest version

### Cloud provider

only k8s

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by olderTaoist** [comment]
not sure if it is a bug. in our gpu cluster, gpu machines are generally marked with taint, and pods that do not use gpu resources will not tolerate taint, or use priority class, and pods that use gpu resources have a higher level. even traversing the pod to use the resources won't solve your case.

**Comment by LY-today** [comment]
> not sure if it is a bug. in our gpu cluster, gpu machines are generally marked with taint, and pods that do not use gpu resources will not tolerate taint, or use priority class, and pods that use gpu resources have a higher level. even traversing the pod to use the resources won't solve your case.

You are right, we also have this deployment model. What my MR describes is that when a pod does not apply for a certain resource, the score of the resource should not be calculated on the pod just because the configuration parameter has the resource. Otherwise it will lead to continued amplification of this strategy.

**Comment by ditingdapeng** [comment]
Amazing! I encountered this issue during testing as well, but I assumed it was an official feature and didn't follow up further. Very impressive find!
